### PR TITLE
webgpu: Replace timestamp-query-in-passes with timestamp-query

### PR DESCRIPTION
**Pull request by gyagp**
Timestamp-query is a core feature, which should be generally supported on all platforms, including macOS. Note that Chrome switch '--enable-dawn-features=allow_unsafe_apis' is still needed now as the timestamp has the accuracy of nanosecond, which is too accurate to be safe. Later changes in Chrome may lift this limitation.

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

### CODE REVIEW

File: tfjs-backend-webgpu/src/base.ts [hunk 1]
```diff
@@ -32,16 +32,7 @@ if (isWebGPUSupported()) {
...
-    // Note that timestamp-query-inside-passes is not formally in spec as
-    // timestamp within a pass is not generally supported on all the platforms.
-    // More details can be found at
-    // https://github.com/gpuweb/gpuweb/blob/main/proposals/timestamp-query-inside-passes.md
-    if (adapter.features.has('timestamp-query-inside-passes')) {
-      deviceDescriptor.requiredFeatures =
-          // tslint:disable-next-line:no-any
-          ['timestamp-query-inside-passes' as any];
-    }
+    deviceDescriptor.requiredFeatures = ['timestamp-query'];
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/base.ts:35]: Do we need to check whether `adapter.features.has('timestamp-query')` is true?

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 2]
```diff
@@ -118,9 +119,12 @@ export class WebGPUBackend extends KernelBackend {
   private pipelineCache:
       {[key: string]: GPUComputePipeline|Promise<GPUComputePipeline>};
   private programTimersStack: TimerNode[];
-  private querySet: GPUQuerySet;
+
+  private queryResolveBuffer: GPUBuffer = null;
+  private querySet: GPUQuerySet = null;
+  private querySetCount = 2;
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/backend_webgpu.ts:125]: `querySetCount` is not needed since it's always 2?

**Review by qjia7 — COMMENTED** [review]
Please also make sure the bots are green.

File: tfjs-backend-webgpu/src/backend_webgpu.ts (hunk 2 above)
Comment (gyagp) [reply tfjs-backend-webgpu/src/backend_webgpu.ts:124]: This needs to be used in 2 functions.

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 3]
```diff
@@ -966,57 +935,78 @@ export class WebGPUBackend extends KernelBackend {
...
-      (pass as any).writeTimestamp(this.querySet, 1);
-    }
-    this.dispatchNumberInEncoder++;
+        queryStagingBuffer = this.bufferManager.acquireBuffer(
+            this.querySetCount * 8,
+            GPUBufferUsage.MAP_READ | GPUBufferUsage.COPY_DST);
+
+        this.commandEncoder.copyBufferToBuffer(
+            this.queryResolveBuffer, 0, queryStagingBuffer, 0,
+            this.querySetCount * 8);
```
[the remark this answers is not in the document]
Comment (gyagp) [reply tfjs-backend-webgpu/src/backend_webgpu.ts]: Act

File: tfjs-backend-webgpu/src/base.ts (hunk 1 above)
Comment (gyagp) [reply tfjs-backend-webgpu/src/base.ts]: yes, we need a check

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 4]
```diff
@@ -660,15 +646,15 @@ export class WebGPUBackend extends KernelBackend {
   }
 
   override async time(f: () => void): Promise<WebGPUTimingInfo> {
-    if (!this.supportTimeQuery) {
+    if (!this.supportTimestampQuery) {
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/backend_webgpu.ts:649]: Is it better to throw the warning once?

File: tfjs-backend-webgpu/src/base.ts [hunk 5]
```diff
@@ -33,14 +33,10 @@ if (isWebGPUSupported()) {
...
-    // More details can be found at
-    // https://github.com/gpuweb/gpuweb/blob/main/proposals/timestamp-query-inside-passes.md
-    if (adapter.features.has('timestamp-query-inside-passes')) {
-      deviceDescriptor.requiredFeatures =
-          // tslint:disable-next-line:no-any
-          ['timestamp-query-inside-passes' as any];
+    if (adapter.features.has('timestamp-query')) {
+      deviceDescriptor.requiredFeatures = ['timestamp-query'];
+    } else {
+      deviceDescriptor.requiredFeatures = [];
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/base.ts:39]: nit: `else` is not needed?

**Review by qjia7 — APPROVED** [review]
LGTM with nits.

**Comment by gyagp** [comment]
To solve the issues in bots, it seems we need to upgrade TypeScript to the latest. I will wait for https://github.com/tensorflow/tfjs/pull/7681.

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 6]
```diff
@@ -966,57 +946,85 @@ export class WebGPUBackend extends KernelBackend {
...
-
-    if (shouldTimeProgram && this.supportTimeQuery) {
-      // tslint:disable-next-line:no-any
-      (pass as any).writeTimestamp(this.querySet, 1);
+    this.dispatchCountInPass++;
+
+    if (shouldTimeProgram ||
+        env().get('WEBGPU_DEFERRED_SUBMIT_BATCH_SIZE') as
+            number <= this.dispatchCountInPass) {
+      this.endComputePassEncoder();
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/backend_webgpu.ts:988]: Move `this.endComputePassEncoder();` to `this.getQueryTime()`?

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 7]
```diff
@@ -966,57 +946,85 @@ export class WebGPUBackend extends KernelBackend {
...
+      if (shouldTimeProgram) {
+        this.activeTimers.push(
+            {name: program.constructor.name, query: this.getQueryTime()});
+      } else {
+        this.submitQueue();
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/backend_webgpu.ts:993]:
```
this.endComputePassEncoder();
this.submitQueue();

File: tfjs-backend-webgpu/src/backend_webgpu.ts (hunk 6 above)
Comment (qjia7) [reply tfjs-backend-webgpu/src/backend_webgpu.ts:988]: Please ignore this comment. I misunderstand the conditions.

**Review by xhcao — APPROVED** [review]
LGTM

**Comment by gyagp** [comment]
@pyu10055 @mattsoulanille @Linchenn Based on this PR, you may measure the WebGPU kernel time on macOS now. Note that Chrome option --disable-dawn-features=disallow_unsafe_apis is still needed.

**Pull request merged**